File: php/src/Advertisements/User/Domain/UserBase.php
<?php
declare(strict_types=1);

namespace Demo\App\Advertisements\User\Domain;

use Demo\App\Advertisements\Shared\ValueObjects\Email;
use Demo\App\Advertisements\Shared\ValueObjects\Password;
use Demo\App\Advertisements\Shared\ValueObjects\UserId;
use Demo\App\Advertisements\User\Domain\ValueObjects\Role;
use Demo\App\Advertisements\User\Domain\ValueObjects\UserStatus;

abstract class UserBase
{
    protected ?Password $password = null;

    protected function __construct(
        protected readonly UserId $id,
        protected Email           $email,
        protected Role            $role,
        protected UserStatus      $status,
    ) {
    }

    public function id(): UserId
    {
        return $this->id;
    }

    public function email(): Email
    {
        return $this->email;
    }

    public function password(): ?Password
    {
        return $this->password;
    }

    public function role(): Role
    {
        return $this->role;
    }

    public function status(): UserStatus
    {
        return $this->status;
    }
}

File: kotlin/src/main/kotlin/advertisements/user/domain/UserBase.kt
package advertisements.user.domain

import advertisements.shared.value_object.Email
import advertisements.shared.value_object.Password
import advertisements.shared.value_object.UserId
import advertisements.user.domain.value_object.Role
import advertisements.user.domain.value_object.UserStatus

abstract class UserBase(
    protected val id: UserId,
    protected val email: Email,
    protected val role: Role,
    protected var status: UserStatus
) {
    protected var password: Password? = null

    fun id(): UserId {
        return id
    }

    fun email(): Email {
        return email
    }

    fun password(): Password? {
        return password
    }

    fun role(): Role {
        return role
    }

    fun status(): UserStatus {
        return status
    }
}